docs/syntax-key-destructuring.md

```diff
@@ -0,0 +1,214 @@
++         | `{` keydestructor { [`,` keydestructor] } [`,`] `}` [':' Type]
+
++keydestructor = `.` NAME [`as` NAME] [':' Type]
+```
+
+This would allow for the following:
+
+```lua
+local { .a, .b }, c = t
+
+for _, { .a, .b } in ts do
+end
+```
+
+In all of these cases, `.x` is an index of the table it corresponds to, assigned to a local variable of that name. For example, `local { .a, .b } = t` is syntax sugar for:
+
+```lua
+local a = t.a
+local b = t.b
+```
+
+This will have the same semantics with regards to `__index`, in the order of the variables being assigned. Furthermore, if `t` for whatever reason cannot be indexed (not a table, nil), you will get the same errors as you would if you were writing out `t.a` by hand.
+
+Trying to use object destructuring in a local assignment without a corresponding assignment, such as...
+```lua
+local { .x, .y }
+```
+
+...is valid to parse and will execute, but can be linted against. This might even be emergent, as this will be statically equivalent to `nil.x`, which Luau's type system can catch.
+
+#### Function arguments
+Functions use `parlist`, which eventually uses `binding`. However, attempting to use key destructuring in a function body is not supported.
+
+```lua
+-- NOT supported
+local function f({ .x, .y })
+```
+
+#### Types
+An optional type can be supplied, such as:
+```lua
+local { .a: string, .b: number } = t
+
+-- Equivalent to...
+local a: string = t.a
+local b: number = t.b
+```
+
+Without explicit types, local assignments and for loops will assume the type of `<rhs>.<field>`. For example...
+```lua
+-- x and y will both be typed `number` here
+local { .x, .y } = position :: { number }
+```
+
+Additionally, you can specify a type on the "table" as a whole.
+
+```lua
+local { .x, .y }: Position = p
+```
+
+Combining both is acceptable, in which case the type on the variable takes priority:
+
+```lua
+-- `x` will be `number`, `y` will be the type of `T.y`
+local { .x: number, .y }: T = p
+```
+
+#### Variable renaming
+
+This proposal allows for renaming the assignments using `as`.
+
+```lua
+local { .real as aliased } = t
+-- is equivalent to...
+local aliased = t.real
+```
+
+This helps support mutliple assignments on the same name:
+```lua
+local { .name as nameA } = getObject(a)
+local { .name as nameB } = getObject(b)
+```
+
+#### Arrays
+This RFC does not concern itself with array destructuring. This is in part because it is not obvious what a reasonable syntax for it would be, and because its purposes might be better suited by tuples.
+
+#### Reassignments
+We do not support key destructuring in reassignments, for example...
+
+```lua
+{ .x, .y } = position
+```
+
+This is to avoid ambiguity with potential table calls:
+
+```lua
+local a = b
+{ .x, .y } = c
+```
+
+While with only this RFC, this is unambiguous with 2 lookahead (`{` + `.`), it could potentially complicate our ability to expand the key destructuring syntax, such as to add renaming variables.
+
+## Drawbacks
```

Comment:
IMO these drawbacks will come up more often than not and raise the cognitive load required to properly use and understand this syntax.

That seems counter to the intent of this proposal which is to make things more simple and convenient to use.